pm/sender.go

```diff
@@ -0,0 +1,82 @@
+package pm
+
+import (
+	"math/big"
+	"sync"
+	"sync/atomic"
+
+	ethcommon "github.com/ethereum/go-ethereum/common"
+	"github.com/livepeer/go-livepeer/eth"
+	"github.com/pkg/errors"
+)
+
+// Sender enables starting multiple probabilistic micropayment sessions with multiple recipients
+// and create tickets that adhere to each session's params and unique nonce requirements.
+type Sender interface {
+	StartSession(recipient ethcommon.Address, ticketParams TicketParams) string
+
+	CreateTicket(sessionID string) (*Ticket, *big.Int, []byte, error)
+
+	// Later: support receiving new recipientRandHash values mid-stream
+}
+
+type session struct {
+	senderNonce uint64
+
+	recipient ethcommon.Address
+
+	ticketParams TicketParams
+}
+
+type defaultSender struct {
+	accountManager eth.AccountManager
+
+	sessions sync.Map
+}
+
+// NewSender creates a new Sender instance.
+func NewSender(accountManager eth.AccountManager) Sender {
+	return &defaultSender{
+		accountManager: accountManager,
+	}
+}
+
+func (s *defaultSender) StartSession(recipient ethcommon.Address, ticketParams TicketParams) string {
+	sessionID := hashToHex(ticketParams.RecipientRandHash)
+
+	s.sessions.Store(sessionID, &session{
+		recipient:    recipient,
+		ticketParams: ticketParams,
+		senderNonce:  0,
+	})
+
+	return sessionID
+}
+
+func (s *defaultSender) CreateTicket(sessionID string) (*Ticket, *big.Int, []byte, error) {
+	recipientRandHash := hexToHash(sessionID)
+
+	tempSession, ok := s.sessions.Load(sessionID)
+	if !ok {
+		return nil, nil, nil, errors.Errorf("cannot create a ticket for an unknown session: %+v", sessionID)
+	}
+	session := tempSession.(*session)
+
+	senderNonce := atomic.AddUint64(&session.senderNonce, 1)
```

Comment:
We could do that by assigning `senderNonce - 1` to the ticket, but I thought that would be more confusing, and didn't really see a problem with nonce starting from 1 and not 0.

Another solution could be to assign `-1` as the first value of `session.senderNonce` but that also felt clunky and confusing.

Do you have a solution in mind that is easy to read and still uses the `atomic` functionality?

Reply:
I guess my initial reaction was why wouldn't `senderNonce` start at 0 (i.e. is there some reason for it not to). However, I don't feel particularly strongly about this and I think starting a nonce at 0 or 1 is ultimately a matter of preference (as you mentioned, I don't think there are positive/negative implications for either relative to the other) so feel free to go with what you're the most comfortable with.

If you did want to start `senderNonce` at 0, could you create the ticket prior to updating `session.senderNonce`, so the ticket created always uses the current nonce value prior to its update?

Reply:
I believe the flow you have in mind could result in a race condition, since more than one goroutine could run the same lines of code in parallel that take the current value of `senderNonce`, say 0, and then when the increment it, one will increment to 1, the other to 2, because we're using atomic increment, but both tickets would have `senderNonce` of zero.

Does this make sense?

Reply:
Ah yeah I think you're right about the race condition.

I'm ok with starting `senderNonce = 1` if that means we can avoid implementing an additional lock.